docs/zkapps/tutorials/06-offchain-storage.mdx

```diff
@@ -0,0 +1,402 @@
+* `MerkleWitness8`: The type of our merkle tree witness. Others are available for input, such as `MerkleWitness32` and `MerkleWitness256`. This is necessary for SnarkyJS to use the same instances of the witness cross-library.
+
+Continuing, let's setup our smart contract:
+
+```ts
+...
+ 21 export class NumberTreeContract extends SmartContract {
+ 22   @state(PublicKey) storageServerPublicKey = State<PublicKey>();
+ 23   @state(Field) storageNumber = State<Field>();
+ 24   @state(Field) storageTreeRoot = State<Field>();
+ 25
+ 26   
+ 27   deploy(args: DeployArgs) {
+ 28     super.deploy(args);
+ 29     this.setPermissions({
+ 30       ...Permissions.default(),
+ 31       editState: Permissions.proofOrSignature(),
+ 32     });
+ 33   }
+ 34
+ 35   @method init(storageServerPublicKey: PublicKey) {
+ 36     this.storageServerPublicKey.set(storageServerPublicKey);
+ 37     this.storageNumber.set(Field(0));
+ 38
+ 39     const emptyTreeRoot = new Experimental.MerkleTree(8).getRoot();
+ 40     this.storageTreeRoot.set(emptyTreeRoot);
+ 41   }
+...
+```
+
+Here we add 3 pieces of state to our contract: the public key of the storage server, the "storageNumber"—which is used to ensure the storage server is actively storing states, and the root of the merkle tree.
+
+We also initialize our zkApp's initialize for these three values, by setting them to the public key of our storage server, setting storage number to 0, and storing the root of an empty tree.
+
+Continuing, here is our update function:
+
+```ts
+...
+ 43   @method update(
+ 44     leafIsEmpty: Bool,
+ 45     oldNum: Field,
+ 46     num: Field,
+ 47     path: MerkleWitness8,
+ 48     storedNewRootNumber: Field,
+ 49     storedNewRootSignature: Signature
+ 50   ) {
+ 51     const storedRoot = this.storageTreeRoot.get();
+ 52     this.storageTreeRoot.assertEquals(storedRoot);
+ 53
+ 54     let storedNumber = this.storageNumber.get();
+ 55     this.storageNumber.assertEquals(storedNumber);
+ 56
+ 57     let storageServerPublicKey = this.storageServerPublicKey.get();
+ 58     this.storageServerPublicKey.assertEquals(storageServerPublicKey);
+ 59
+ 60     let leaf = [oldNum];
+ 61     let newLeaf = [num];
+ 62
+ 63     // newLeaf can be a function of the existing leaf
+ 64     newLeaf[0].assertGt(leaf[0]);
+ 65
+ 66     const updates = [
+ 67       {
+ 68         leaf,
+ 69         leafIsEmpty,
+ 70         newLeaf,
+ 71         newLeafIsEmpty: Bool(false),
+ 72         leafWitness: path,
+ 73       },
+ 74     ];
+ 75
+ 76     const storedNewRoot = OffChainStorage.assertRootUpdateValid(
+ 77       storageServerPublicKey,
+ 78       storedNumber,
+ 79       storedRoot,
+ 80       updates,
+ 81       storedNewRootNumber,
+ 82       storedNewRootSignature
+ 83     );
+ 84
+ 85     this.storageTreeRoot.set(storedNewRoot);
+ 86     this.storageNumber.set(storedNewRootNumber);
+ 87   }
+ 88 }
+```
+
+We get and assert the current state of the contract - and then we perform the update.
```

Comment:
@es92 Is `OffChainStorage.assertRootUpdateValid` also updating the data in the storage? Or is that step skipped?

Reply:
When you say, "updating the data in storage", do you mean updating the data at the remote storage server, or updating the merkle root of the contract?

Its doing neither actually - its checking that the updates (of which assertions have been made above), actually link the existing merkle root in the repo, to a new merkle root that is being stored.

Lmk if that helps - and if so, perhaps if it can be explained better - I was attempting to explain all that is happening in the section above, `Notice we import some items from zkapp-offchain-storage: ...`, but sounds like that could be missing part of the needed explanation.

Reply:
Going to move towards merigng per discussion on slack - think would be great to address eventually though, I can imagine this being confusing. I'll add more discussion on this in the text for now too